### PR TITLE
[JUJU-875] Remove use of "Force=true" when removing an application

### PR DESCRIPTION
**Pull request by benhoyt**
Remove "Force=true" in garbageCollectOneApplication when removing an application and clean up properly.

We believe we had originally added this "destroyOp.Force = true" to get something going quickly initially, but it's not appropriate to force when the user hasn't requested it.

Once we removed this we also had to do the following so the units and application get removed correctly:

* Clear the "has-resources" flag once the K8s resources have been deleted. This is done in the caasapplicationprovisioner application worker's dead() method.
* Also in caasapplicationprovisioner, watch the application's units and call RemoveUnit when a unit is dead (for pod-spec charms, this is done in worker/caasoperator, but we weren't doing it for sidecar charms).

I've added unit tests for all the new code paths, plus  a basic integration test for `juju remove-application --force`, and we already have an integration test for `remove-application` without force (https://github.com/juju/juju/pull/13929).

The commit https://github.com/juju/juju/pull/13973/commits/5ebc199b2de89872ad053f2a82d46b0447a5faf0 should also fix the race condition that causes an application not to be removed from state sometimes: https://bugs.launchpad.net/juju/+bug/1970508

## QA steps

See integration tests -- just `deploy` and `remove-application` on a sidecar application:

```
juju bootstrap microk8s
# optional logging for debugging:
# juju model-config logging-config="<root>=INFO;juju.worker.caasapplicationprovisioner=DEBUG;juju.state=DEBUG"
juju deploy snappass-test
# wait for it to deploy
juju remove-application snappass-test
juju status
```

To repro the shutdown race condition, do the following:

```
# bootstrap and add logging as per above
juju deploy snappass-test -n10
# wait for it to deploy
juju remove-application snappass-test
juju status
```

Every other time or so with 10 units it will not delete the application from state due to the race condition mentioned in https://github.com/juju/juju/pull/13973/commits/5ebc199b2de89872ad053f2a82d46b0447a5faf0.

### CODE REVIEW

File: api/controller/caasapplicationprovisioner/client.go [hunk 1]
```diff
@@ -22,6 +19,8 @@ import (
 	"github.com/juju/juju/docker"
 	"github.com/juju/juju/rpc/params"
 	"github.com/juju/juju/storage"
+	"github.com/juju/names/v4"
+	"github.com/juju/version/v2"
```
Comment (wallyworld) [inline api/controller/caasapplicationprovisioner/client.go:23]: move me

File: state/application.go [hunk 2]
```diff
@@ -2169,11 +2169,11 @@ func (a *Application) ClearResources() error {
 				{"has-resources", true}},
 			Update: bson.D{{"$set", bson.D{{"has-resources", false}}}},
 		}}
-		logger.Debugf("application %q still has cluster resources, scheduling cleanup", a.doc.Name)
+		logger.Debugf("application %q now has no cluster resources, scheduling cleanup", a.doc.Name)
```
Comment (wallyworld) [inline state/application.go:2172]: oh dear, thanks for fixing comment

**Review by wallyworld — COMMENTED** [review]
It seems we've moved some of the core facade methods across from the operator provisioner, but is the application undertaker also implemented for sidecar charms? The undertaker is the worker that monitors the cluster for resource removal. Is it the case that the snappass remove test works because there are no resources created? Would a charm which does create resources have its app removed while resources remain?

File: api/controller/caasapplicationprovisioner/client.go [hunk 3]
```diff
@@ -86,14 +85,20 @@ func maybeNotFound(err *params.Error) error {
 	return errors.NewNotFound(err, "")
 }
 
-// Life returns the lifecycle state for the specified CAAS application
+// Life returns the lifecycle state for the specified application
```
Comment (jameinel) [inline api/controller/caasapplicationprovisioner/client.go:88]: no longer an 'application' according to your switch statement

Comment (benhoyt) [reply api/controller/caasapplicationprovisioner/client.go:88]: Actually, the comment was:

```
// Life returns the lifecycle state for the specified CAAS application
// or unit in the current model.
```

The "or unit" was there (presumably a copy'n'paste error from earlier?) even though the function didn't actually work on units. Now it does. I just removed the "CAAS" part as it wasn't actually CAAS-specific.

**Comment by benhoyt** [comment]
Ian, per our discussion, the application undertaker for pod-spec charms waits for workload pods and operator pods to disappear, then clear the has-resources flag. It's a bit different here for sidecar charms: the worker code in `dead()` calls [`waitForTerminated`](https://github.com/juju/juju/blob/2f4d48cf0f4df721fa23ddd433b96a241c5b3ffb/worker/caasapplicationprovisioner/application.go#L823) which calls `app.Exists()` and waits for it to all go away in K8s before (now) clearing the has-resources flag.

**Comment by benhoyt** [comment]
$$merge$$

**Pull request merged**